packages/node/src/ethereum/utils/free-balance-app.ts

```diff
@@ -59,31 +54,22 @@ export function flip(coinTransferMap: CoinTransferMap): CoinTransferMap {
 /**
  * Returns the first base mapping, but incremented by values specified in the
  * second increment. Passing increments whose keys are not present in the base
- * is an error. Keys in the base mapping which are not explicitly incremented
- * are returned unchanged.
+ * sets them to the increment. Keys in the base mapping which are not explicitly
+ * incremented are returned unchanged.
  */
 export function merge(
   base: { [s: string]: BigNumber },
   increments: { [s: string]: BigNumber }
 ) {
   const ret = {} as { [s: string]: BigNumber };
 
-  for (const key of Object.keys(base)) {
-    if (increments[key]) {
-      ret[key] = base[key].add(increments[key]);
-      if (ret[key].lt(Zero)) {
-        throw new Error("Underflow in merge");
-      }
-    } else {
-      ret[key] = base[key];
-    }
-  }
+  const s1 = new Set(Object.keys(base));
+  const s2 = new Set(Object.keys(increments));
 
-  for (const key of Object.keys(increments)) {
-    if (!base[key]) {
-      throw new Error(
-        `mismatch: ${Object.keys(increments)} ⊄ ${Object.keys(base)}`
-      );
+  for (const key of new Set([...s1, ...s2])) {
+    ret[key] = (base[key] || Zero).add(increments[key] || Zero);
+    if (ret[key].lt(Zero)) {
+      throw new Error("Underflow in merge");
```

Comment:
Since this error is very specific to this context, can it be more descriptive about said context?

i.e. if this is thrown, a Node operator is unsure of what is being merged and furthermore is unsure of where to start with investigating an "underflow" (another term that could use updating, maybe mentioning relative to 0 or something?)

Reply:
moved to caller